source/_posts/2017-07-22-robot-framework-as-embedded-systems-validation-framework.markdown

```diff
@@ -0,0 +1,350 @@
+    Read Until    iPXE> \x1b[?25h
+
+Download and boot pxelinux
+    # request IP address
+    Write Bare Slow  dhcp net0\n
+    Read Until    ok
+    Read Until    iPXE>
+    # provide pxelinux filename on PXE server
+    Write Bare Slow  set filename pxelinux.0\n
+    Read Until    iPXE>
+    # provide PXE server IP address
+    Write Bare Slow  set next-server %{PXE_SRV_IP}\n
+    Read Until    iPXE>
+    # download and boot pxelinux
+    Write Bare Slow  chain tftp://\${next-server}/\${filename}\n
+    Read Until    PXE server boot menu
+    Close Connection
+```
+
+Configuration for `ser2net` port and PXE server IP address are passed through
+environment variables.
+
+## Automated PXE booting
+
+Assuming your PXE server works fine you can run:
+
+```
+git clone https://github.com/pcengines/apu-test-suite.git
+cd apu-test-suite
+sudo ser2net -c ser2net_apu.cfg
+```
+
+Then please change port and IP address accordingly:
+
+```
+S2N_PORT=13542 PXE_SRV_IP=<IP_ADDR> robot -b debug.log -L TRACE pxe_boot.robot
+```
+
+Please note that port is hardcoded in ser2net_apu.cfg.
+
+Output in terminal should look like this:
+
+```
+==============================================================================
+Pxe Boot                                                                      
+==============================================================================
+Enter iPXE shell                                                      | PASS |
+------------------------------------------------------------------------------
+Download and boot pxelinux                                            | PASS |
+------------------------------------------------------------------------------
+Pxe Boot                                                              | PASS |
+2 critical tests, 2 passed, 0 failed
+2 tests total, 2 passed, 0 failed
+==============================================================================
+Debug:   /home/pietrushnic/storage/wdc/projects/2017/pcengines/apu/src/apu-test-suite/debug.log
+Output:  /home/pietrushnic/storage/wdc/projects/2017/pcengines/apu/src/apu-test-suite/output.xml
+Log:     /home/pietrushnic/storage/wdc/projects/2017/pcengines/apu/src/apu-test-suite/log.html
+Report:  /home/pietrushnic/storage/wdc/projects/2017/pcengines/apu/src/apu-test-suite/report.html
+```
+
+This means PXE boot menu is already in your telnet, to connect to it simply
+type:
+
+```
+telnet localhost 13542
+```
+
+After refreshing your screen with `<CTRL>-L` you should see boot menu:
+
+<a class="fancybox" rel="group" href="/assets/images/pxe_server_menu.png"><img src="/assets/images/pxe_server_menu.png" alt="" /></a>
+
+Moreover you can investigate in details what happened using automatically
+generated HTML page:
+
+<a class="fancybox" rel="group" href="/assets/images/ipxe_test_log.png"><img src="/assets/images/ipxe_test_log.png" alt="" /></a>
+
+## Summary
+
+Why bother ? Firmware debugging effort consists of tons of repeatable tasks.
```

Comment:
Is the spacing between `bother` and `?` formatting mistake or rather an intentional operation?